SegmentedView for iOS and Android

hnabbasi opened the pull request:
### Description of Change ###

New control added from #385 

>Note: Migrating from Xamarin.Forms [Issue 6904](https://github.com/xamarin/Xamarin.Forms/issues/6904)


Have a segmented control for Xamarin.Forms. iOS has `UISegmentedControl`, but Android does not. This control will provide native looking segments for Android platform.

iOS | Android
--|--
<img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/Segments_iOS_Text.gif"/> | <img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/Segments%20Android%20Text.gif"/>
<img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/Segments_iOS_Image.gif"/> |<img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/Segments%20Android%20Image.gif" />
|| <img src="https://raw.githubusercontent.com/hnabbasi/Segments/master/media/Segments_Corner_Radius.gif" />


## API
The `SegmentedView` implements `IList<string>` with `DisplayMode` of `Text` or `Image`.

`SegmentedView`:
```csharp
class SegmentedView : View {}
```

## Properties
### SegmentedView

Name | Type | Description
---|---|---
Display Mode | Enum | Display mode is `SegmentMode.Text` by default and can be set to `SegmentMode.Image` as an alternative.
Color | Xamarin.Forms.Color | Main color for the control
Items | IList\<string\> | List of string values for title or image source
ItemsSource | IList | Items source for the values
SelectedItem | object | Selected item   
SelectedIndexChanged | Event | Event handler for when a segment is selected. Raises `SelectedItemChangedEventArgs` with segment and selected index
SelectedIndex | System.Int32 | Index of selected segment

### Platform Specific - Android
Name | Type | Description
---|---|---
CornerRadius | CornerRadius | Corner radius on Android ONLY

## Usage

**XAML**
```xml
<SegmentedView SelectedIndexChanged="OnSegmentSelected">
    <SegmentedView.Items>
        <x:String>View A</x:String>
        <x:String>View B</x:String>
    </SegmentedView.Items>
</SegmentedView>

<!-- Using ItemsSource -->

<SegmentedView 
    SelectedIndexChanged="OnSegmentSelected"
    ItemsSource="{Binding values}">
</SegmentedView>
```
**C#**
```csharp
var segments = new SegmentedView
    {
        Children = new List<string>
        {
            "View A",
            "View B",
            "View C"
        }
    };
segments.SelectedIndexChanged += OnSegmentSelected;

// handler
void OnSegmentSelected(object sender, SelectedItemChangedEventArgs e){ }
```

## Platforms
### iOS
On iOS, this is simply a `UISegmentedControl`.

### Android
On Android, this is be a `RadioGroup` with a custom `RadioButton` that renders a simple text or bitmap drawable in image mode.

```csharp
class FormsSegmentsView : Android.Widget.RadioGroup
```

### UWP
👀 TBD 

### PR Checklist ###
<!-- Please check all the things you did here and double-check that you got it all, or state why you didn't do something -->

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard
<!-- If at all possible, please update/add the documentation on the repo below. We would very much appreciate that. If you are unable to, please consider at least opening an issue on the repo below so we know that Docs still need to be adjusted/created. Thanks! <3 -->
- [ ] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

hnabbasi commented:
On Android, CornerRadius get some love...

## Usage
In XAML,
```xml
<SegmentedView
    ...
    CornerRadius="20" />
```
Or individual configs: TopLeft, TopRight, BottomRight, BottomLeft
```xml
<SegmentedView
    ...
    CornerRadius="20,0,20,0" />
```

![](https://raw.githubusercontent.com/hnabbasi/Segments/master/media/SegmentsAndroid.gif)

pictos commented:
@hnabbasi can you resolve the conflicts?

pictos commented on XamarinCommunityToolkit/Extensions/iOS/ImageResourceExtension.ios.cs:
```diff
@@ -0,0 +1,48 @@
+
+			try
+			{
+#if __MOBILE__
```
This file doesn't compile to MacOS, can you remove this?

hnabbasi replied:
Renamed the file to compile for MacOS.

pictos commented on XamarinCommunityToolkit/Extensions/iOS/ImageResourceExtension.ios.cs:
```diff
@@ -0,0 +1,48 @@
+			if (source == null || source.IsEmpty)
+				return null;
+
+			var handler = Forms.Internals.Registrar.Registered.GetHandlerForObject<IImageSourceHandler>(source);
```
Can you check if the linker will remove this?

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/Android/FormsSegments.android.cs:
```diff
@@ -0,0 +1,294 @@
+			SegmentSelected?.Invoke(this, new SelectedPositionChangedEventArgs(IndexOfChild(CurrentSegment)));
+		}
+
+		#region Drawable Resources
```
Can you remove this `#region` and others that may have in this PR?

hnabbasi replied:
Done.

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/Android/FormsSegments.android.cs:
```diff
@@ -0,0 +1,294 @@
+				}
+			});
+
+			return tcs.Task.Result;
```
Can this lead a `dead lock`?

hnabbasi replied:
Since I am using `TaskCompletionSource` and don't want the `async` identifier, not sure how to get the result without doing `.Result`. Any ideas?

roubachof replied:
I don't really get why you use `TaskCompletionSource` here. Normally it is used to make sync code async.
And you could have the same semantic doing:

```csharp
var task = Task.Run(async () => 
{
    var formsBitmap = await context.GetFormsBitmapAsync(filePath));
    return new BitmapDrawable(Resources, formsBitmap);
});

return task.GetAwaiter().GetResult();
```

Which is still bad as @pictos pointed it cause it could lead to deadlock or thread pool starvation.
I guess you should make your method async and adapt your code hierarchy accordingly: when it's async, it's async all the way :)

pictos replied:
Yeah, the @roubachof 's suggestion is valid, but since this code will run in a program that has an UIThread, this can be dangerous. So do we need this code finish to move to the next line or this code can run in parallel? If the logic need this peace to get done to do the next step, we need to rethink all the API and make it async friendly

roubachof replied:
Nope won't be dangerous since the Synchronization context will be restored after the await statement. Also the iOS version has the same issue, just need to really go async till the root of the code :)

hnabbasi replied:
Well, I was trying to avoid `async` chain due to a single `Task` and run it inside a synchronous method. But I see that it may cause an issue at some point. I am reviewing the API now. Great feedback, keep it coming.

pictos replied:
@hnabbasi if you need any help or want to discuss more, let us know.

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/Android/FormsSegments.android.cs:
```diff
@@ -0,0 +1,294 @@
+			disposed = true;
+			if (disposing)
+			{
+				Children.CollectionChanged -= CollectionChanged;
```
Can we check for null?

hnabbasi replied:
Done.

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/Android/SegmentedViewRenderer.android.cs:
```diff
@@ -0,0 +1,139 @@
+{
+	public class SegmentedViewRenderer : ViewRenderer<SegmentedView, FormsSegments>
+	{
+		readonly Context context;
```
The `ViewRenderer` exposes the `Context` as property, can you remove this?

hnabbasi replied:
Done

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/Android/SegmentedViewRenderer.android.cs:
```diff
@@ -0,0 +1,139 @@
+	public class SegmentedViewRenderer : ViewRenderer<SegmentedView, FormsSegments>
+	{
+		readonly Context context;
+		FormsSegments control;
```
Can you rename this, I want to avoid confusion with the `Control` prop. I know, the IDE will help me, but reviewing here on the GitHub website can be hard.

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/Android/SegmentedViewRenderer.android.cs:
```diff
@@ -0,0 +1,139 @@
+				return;
+
+			if (e.PropertyName == SegmentedView.SelectedIndexProperty.PropertyName)
+			{
```
we don't use `{ }` for one line code. Can you remove it and others, as you did in line 118?

hnabbasi replied:
Done.

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/SegmentedView.shared.cs:
```diff
@@ -0,0 +1,199 @@
+			BindableProperty.Create(nameof(SelectedItem), typeof(object), typeof(SegmentedView), null, BindingMode.TwoWay,
+									propertyChanged: OnSelectedItemChanged);
+
+		public IList ItemsSource
```
Normally we use `IEnumerable` as ItemSource property, to make it works with a large range of Collections, any reason to use `IList`?

hnabbasi replied:
I am getting the count of the `IList` when drawing the controls directly from this collection.

roubachof replied:
yup IEnumerable would be best for consistency, use Linq to Count() the items.

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/SegmentedView.shared.cs:
```diff
@@ -0,0 +1,199 @@
+			}
+		}
+
+		static readonly BindableProperty displayProperty =
```
Typo (;
```suggestion
		static readonly BindableProperty DisplayProperty =
```

hnabbasi replied:
Ah. I am not sure what to do with this yet. But sure, we can fix the name 👍

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/SegmentedView.shared.cs:
```diff
@@ -0,0 +1,199 @@
+
+		void OnItemsSourceChanged(IList oldValue, IList newValue)
+		{
+			var oldObservable = oldValue as INotifyCollectionChanged;
+			if (oldObservable != null)
+				oldObservable.CollectionChanged -= CollectionChanged;
```
Can you apply this change in others?

```suggestion
			if (oldObservable is INotifyCollectionChanged oldObservable)
				oldObservable.CollectionChanged -= CollectionChanged;
```

hnabbasi replied:
Done.

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/SegmentedView.shared.cs:
```diff
@@ -0,0 +1,199 @@
+			if (ItemDisplayBinding == null)
+				return item.ToString();
+
+			//TODO: Fix this
+			//ItemDisplayBinding.Apply(item, this, displayProperty);
+			//ItemDisplayBinding.Unapply();
```
Can you remove these comments? Is this fixed or something that we need to care about?

hnabbasi replied:
I need help with this. We don't have `.Apply` and `.Unapply` outside of `Xamarin.Forms` assembly and I was using these to get `ItemDisplayBinding` property to show (as in Picker). Not sure how to get this working. It is a "nice to have" but not a "must have". So, removing for now.

pictos replied:
@hnabbasi please open an issue after we merge this, that way we don't lose this point.

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/SegmentedView.shared.cs:
```diff
@@ -0,0 +1,199 @@
+				case NotifyCollectionChangedAction.Remove:
+					RemoveItems(e);
+					break;
+				default: //Move, Replace, Reset
```
```suggestion
				default:
```

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/SegmentedView.shared.cs:
```diff
@@ -0,0 +1,199 @@
+		{
+			if (!(bindable is SegmentedView segment))
+				return;
+			int.TryParse(newValue?.ToString(), out var index);
```
and if the `TryParse` fails?

hnabbasi replied:
Fixed it. Defaulting to `0` in case of failure.

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/iOS/SegmentedViewRenderer.ios.cs:
```diff
@@ -0,0 +1,169 @@
+
+namespace Xamarin.CommunityToolkit.iOS.UI.Views
+{
+	[Xamarin.Forms.Internals.Preserve(AllMembers = true)]
```
Can you remove this? We want to make this lib linker safe.

hnabbasi replied:
Done.

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/iOS/SegmentedViewRenderer.ios.cs:
```diff
@@ -0,0 +1,169 @@
+					if (img != null)
+						Control.InsertSegment(img, position, false);
+					else
+						Console.WriteLine("ImageSource is null");
```
I think that we use a Log here, to make it more visible to devs. Can you use the `Log` as you did in line 163?

pictos commented on XamarinCommunityToolkit/Views/SegmentedView/iOS/SegmentedViewRenderer.ios.cs:
```diff
@@ -0,0 +1,169 @@
+				UpdateSelectedSegment(Element.SelectedIndex);
+
+			if (e.PropertyName == SegmentedView.ColorProperty.PropertyName)
+			{
```
Can you remove the `{ }` as you did in line 123?

hnabbasi replied:
Oopss. Done.

pictos commented on XamarinCommunityToolkitSample.iOS/Xamarin.CommunityToolkit.Sample.iOS.csproj:
```diff
@@ -139,7 +139,7 @@
   </ItemGroup>
   <ItemGroup>
-    <PackageReference Include="Xamarin.Forms" Version="4.8.0.1269" />
+    <PackageReference Include="Xamarin.Forms" Version="4.8.0.1560" />
```
Any reason to update the package?

hnabbasi replied:
Must have been updated with somehow.

pictos reviewed (changes requested):
@hnabbasi Thanks for this huge contribution! I didn't test it yet, just take a look at the code. If you have any question or needs help please let me know (:

Seuleuzeuh commented:
@hnabbasi Great work !
Does this segementedView is scrollable (when using a big list of items) ?

roubachof commented:
@Seuleuzeuh don't think so. If you want really customizable tabs (including radius), you should use the TabView.

hnabbasi commented:
> @hnabbasi Great work !
> Does this segementedView is scrollable (when using a big list of items) ?

No. You may want to use TabView for that.

domagojmedo commented:
What is the difference between this and TabView?

VladislavAntonyuk commented:
> What is the difference between this and TabView?

I see SegmentedView as a nice picker (choose from multiple options), like color picker
TabView is like a tabpage but view.
imaging SegmentedView as a tabview without content, only tabbar

roubachof commented on XamarinCommunityToolkit/Views/SegmentedView/iOS/SegmentedViewRenderer.ios.cs:
```diff
@@ -0,0 +1,166 @@
+
+		void InitializeControl()
+		{
+			PopulateSegments(Element.Items);
```
There can be a race condition on PopulateSegments since InsertSegment is async void and won't wait until all segments are inserted. My advise would be async all the way x)
so InitializeControl would be async, just like it would be in the android side.

vniehues commented:
Is there a timeframe for when we can expect this to be merged? - I would love to use this in my current projekt. :)

hnabbasi commented:
> Is there a timeframe for when we can expect this to be merged? - I would love to use this in my current projekt. :)

I have addressed all the feedback items but working on fixing a fix for a BackgroundColor issue on Android. Will push out another commit this week. So, it should be soon 😃

jfversluis commented:
I do want to temper the expectations here a little bit. This _might_ be cut for the first stable version. With merging the C# Markup PR we're now feature complete for a first stable package and I would want to focus on stability rather than adding more new shiny things in there. I totally understand we want to have all the things in as soon as possible, but I think in the long run it's better to focus on polishing what is there right now first and start adding things again from there.

I'm not saying this is set in stone and what the timeframe for adding it will be. Or, maybe there is enough time to still add this in, who knows. But just wanted to drop in and manage expectations for a bit.

Having all that said, thank you for all your work and patience working with us!

pictos commented:
@hnabbasi we merged a huge change today, are you able to fix the conflicts?

hnabbasi commented:
> @hnabbasi we merged a huge change today, are you able to fix the conflicts?

I did before today's merge. Will pull again and resolve if anything breaks again

AndreiMisiukevich commented:
> > @hnabbasi we merged a huge change today, are you able to fix the conflicts?
> 
> I did before today's merge. Will pull again and resolve if anything breaks again

Hi :) Are you still interested in bringing this enhancement?)

hnabbasi commented:
> > > @hnabbasi we merged a huge change today, are you able to fix the conflicts?
> 
> > 
> 
> > I did before today's merge. Will pull again and resolve if anything breaks again
> 
> 
> 
> Hi :) Are you still interested in bringing this enhancement?)

Absolutely. I will find some time this weekend 🙂

vniehues commented:
Hey @hnabbasi ! 🙂
Are you still working on this?

hnabbasi commented:
Yes. Just came back from vacation. Will resume. 😊

jfversluis commented:
Great! Sorry for the slow response on this and everything. We'll do better this time around!

Could you make sure to target the develop branch when you get to this? :) Thanks!

Cfun1 commented:
Could you please properly linked this pr to your issue by using the proper keywords [GitHub:linking-a-pull-request-to-an-issue](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue)

acaliaro commented:
Hi guys, there is an ETA for this?

jfversluis commented:
@acaliaro not that I'm aware of. I see @hnabbasi has marked it as a draft so as soon as he is ready (and pings me so I will prioritize the review) it may end up in 1.1.

For the time being however you should be able to use the TabView for this. [Create items with no view](https://github.com/xamarin/XamarinCommunityToolkit/blob/main/samples/XCT.Sample/Pages/Views/TabView/NoContentPage.xaml) and handle the navigation yourself of style the TabView to look like this :)

hnabbasi commented:
New PR is ready to review at #867 

closing this one.

cc: @jfversluis @AndreiMisiukevich @pictos
